### PR TITLE
CASSANDRA-17527: Clients using JMX are unable to handle non-standard java types but we leak this into our interfaces

### PR DESCRIPTION
**Pull request by dcapwell**

### CODE REVIEW

File: src/java/org/apache/cassandra/db/DisallowedDirectories.java [hunk 1]
```diff
@@ -49,14 +50,21 @@ private DisallowedDirectories()
         MBeanWrapper.instance.registerMBean(this, MBEAN_NAME, MBeanWrapper.OnException.LOG);
     }
 
-    public Set<File> getUnreadableDirectories()
+    @Override
+    public Set<java.io.File> getUnreadableDirectories()
```
Comment (dcapwell) [inline src/java/org/apache/cassandra/db/DisallowedDirectories.java:54]: cc @belliottsmith , had to revert the change done in Paxos as it breaks JMX.  I don't see anything using this methods so they are JMX only, so is there any issue with using java's file type?

File: test/unit/org/apache/cassandra/db/guardrails/GuardrailConsistencyLevelsTester.java [hunk 2]
```diff
@@ -24,6 +24,7 @@
 import java.util.function.BiConsumer;
 import java.util.function.Function;
 import java.util.stream.Collectors;
+import java.util.stream.Stream;
 
 import com.google.common.collect.ImmutableSet;
```
Comment (adelapena) [inline test/unit/org/apache/cassandra/db/guardrails/GuardrailConsistencyLevelsTester.java]: Nit: the import of `ImmutableSet` is now unused

File: src/java/org/apache/cassandra/db/ColumnFamilyStoreMBean.java [hunk 3]
```diff
@@ -53,6 +54,7 @@
     /**
      * force a major compaction of specified key range in this column family
      */
+    @BreaksJMX("This API was released in 3.10 using a parameter that takes Range of Token, which can only be done IFF client has Cassandra binaries in the classpath")
```
Comment (adelapena) [inline src/java/org/apache/cassandra/db/ColumnFamilyStoreMBean.java:57]: Shouldn't we provide a JMX-friendly variant of this method?

File: src/java/org/apache/cassandra/service/StorageServiceMBean.java [hunk 4]
```diff
@@ -828,18 +829,22 @@ public StageConcurrency(int corePoolSize, int maximumPoolSize)
...
     @Deprecated
     public void enableAuditLog(String loggerName, Map<String, String> parameters, String includedKeyspaces, String excludedKeyspaces, String includedCategories, String excludedCategories,
                                String includedUsers, String excludedUsers) throws ConfigurationException, IllegalStateException;
 
+    @BreaksJMX("This API was exposed as throwing ConfigurationException, removing is binary compatible but not source; see https://docs.oracle.com/javase/specs/jls/se7/html/jls-13.html")
     @Deprecated
     public void enableAuditLog(String loggerName, String includedKeyspaces, String excludedKeyspaces, String includedCategories, String excludedCategories,
                                String includedUsers, String excludedUsers) throws ConfigurationException, IllegalStateException;
 
+    @BreaksJMX("This API was exposed as throwing ConfigurationException, removing is binary compatible but not source; see https://docs.oracle.com/javase/specs/jls/se7/html/jls-13.html")
```
Comment (adelapena) [inline src/java/org/apache/cassandra/service/StorageServiceMBean.java]: The two variants of `enableAuditLog` that are not marked as deprecated are new to 4.1, shouldn't we make them JMX-friendly instead of marking them with the new annotation?

File: test/unit/org/apache/cassandra/tools/JMXStandardsTest.java [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+{
+    /**
+     * JMX typlically works well with java.* and javax.*, but not all types are serializable and will work, so this class
+     * goes with a explicit approval list, new APIs may fail if a java.* or javax.* is used not in this allow list, if
+     * that is the case it is fine to add here.
+     *
+     * It is never fine to allow non java.* and javax.* types, they can not be handled by clients, so should never be
+     * allowed.
+     */
+    private static Set<Class<?>> ALLOWED_TYPES = ImmutableSet.<Class<?>>builder()
```
Comment (adelapena) [inline test/unit/org/apache/cassandra/tools/JMXStandardsTest.java]: Nit: could be `final`

File: test/unit/org/apache/cassandra/tools/JMXStandardsTest.java [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+                                                 .add(IllegalStateException.class)
+                                                 .add(ClassNotFoundException.class)
+                                                 .add(OpenDataException.class)
+                                                 .build();
+    /**
+     * This list is a set of types under java.* and javax.*, but are too vague that could cause issues; this does not
+     * mean issues will happen with JMX, only that issues may happen only after running and can not be detected at
+     * compile time.
+     */
+    private static Set<Class<?>> DANGEROUS_TYPES = ImmutableSet.<Class<?>>builder()
```
Comment (adelapena) [inline test/unit/org/apache/cassandra/tools/JMXStandardsTest.java]: Nit: could be `final`

File: test/unit/org/apache/cassandra/tools/JMXStandardsTest.java [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+                {
+                    Method method = methods[i];
+                    checkType(method, "return", method.getGenericReturnType(), warnings, errors);
+                    Stream.of(method.getGenericParameterTypes()).forEach(t -> checkType(method, "parameter", t, warnings, errors));
+                    Stream.of(method.getGenericExceptionTypes()).forEach(t -> checkType(method, "throws", t, warnings, errors));
+                }
+            }
+        }
+        if (!warnings.isEmpty())
+            System.err.println(String.join("\n", warnings));
```
Comment (adelapena) [inline test/unit/org/apache/cassandra/tools/JMXStandardsTest.java]: Why don't we produce multiple error messages? Shouldn't we emit log warnings instead?

File: src/java/org/apache/cassandra/db/DisallowedDirectories.java (hunk 1 above)
Comment (dcapwell) [reply src/java/org/apache/cassandra/db/DisallowedDirectories.java:54]: Spoke in Slack, he was cool with me allowing java file

Comment (belliottsmith) [reply src/java/org/apache/cassandra/db/DisallowedDirectories.java:54]: yes, this was a mistake made in CEP-10 when `java.io.File` was replaced - it should have been excluded from the change.

File: src/java/org/apache/cassandra/db/ColumnFamilyStoreMBean.java (hunk 3 above)
Comment (dcapwell) [reply src/java/org/apache/cassandra/db/ColumnFamilyStoreMBean.java:57]: sent out an email on the dev list to figure out what is best; waiting on that to touch these methods, so far only @ekaterinadimitrova2 has replied, and her feedback matches yours (fork)

File: src/java/org/apache/cassandra/service/StorageServiceMBean.java (hunk 4 above)
Comment (dcapwell) [reply src/java/org/apache/cassandra/service/StorageServiceMBean.java]: oh, I might have missed that when looking at git history, prob got confused with the deprecated versions!  Ill double check and fix

File: test/unit/org/apache/cassandra/tools/JMXStandardsTest.java (hunk 7 above)
Comment (dcapwell) [reply test/unit/org/apache/cassandra/tools/JMXStandardsTest.java]: Replacing with

```
warnings.forEach(logger::warn);
```

File: src/java/org/apache/cassandra/service/StorageServiceMBean.java (hunk 4 above)
Comment (dcapwell) [reply src/java/org/apache/cassandra/service/StorageServiceMBean.java]: Confirmed, they were added in CASSANDRA-16725

```
commit b286639eea11b5f6ee709711d00d36aa029bc114
Author: Stefan Miklosovic <smiklosovic@apache.org>
Date:   Wed Jul 28 14:57:02 2021 +0200

    implement getauditlog command

    patch by Stefan Miklosovic; reviewed by Ekaterina Dimitrova, Mick Semb Wever for CASSANDRA-16725

diff --git a/src/java/org/apache/cassandra/service/StorageServiceMBean.java b/src/java/org/apache/cassandra/service/StorageServiceMBean.java
index a5a66073a5..4622d06306 100644
--- a/src/java/org/apache/cassandra/service/StorageServiceMBean.java
+++ b/src/java/org/apache/cassandra/service/StorageServiceMBean.java
@@ -790,8 +790,22 @@ public interface StorageServiceMBean extends NotificationEmitter
     /** Clears the history of clients that have connected in the past **/
     void clearConnectionHistory();
     public void disableAuditLog();
-    public void enableAuditLog(String loggerName, Map<String, String> parameters, String includedKeyspaces, String excludedKeyspaces, String includedCategories, String excludedCategories, String includedUsers, String excludedUsers) throws ConfigurationException;
-    public void enableAuditLog(String loggerName, String includedKeyspaces, String excludedKeyspaces, String includedCategories, String excludedCategories, String includedUsers, String excludedUsers) throws ConfigurationException;
+    public void enableAuditLog(String loggerName, Map<String, String> parameters, String includedKeyspaces, String excludedKeyspaces, String includedCategories, String excludedCategories,
+                               String includedUsers, String excludedUsers, Integer maxArchiveRetries, Boolean block, String rollCycle,
+                               Long maxLogSize, Integer maxQueueWeight, String archiveCommand) throws ConfigurationException, IllegalStateException;
+
+    @Deprecated
+    public void enableAuditLog(String loggerName, Map<String, String> parameters, String includedKeyspaces, String excludedKeyspaces, String includedCategories, String excludedCategories,
+                               String includedUsers, String excludedUsers) throws ConfigurationException, IllegalStateException;
+
+    @Deprecated
+    public void enableAuditLog(String loggerName, String includedKeyspaces, String excludedKeyspaces, String includedCategories, String excludedCategories,
+                               String includedUsers, String excludedUsers) throws ConfigurationException, IllegalStateException;
+
+    public void enableAuditLog(String loggerName, String includedKeyspaces, String excludedKeyspaces, String includedCategories, String excludedCategories,
+                               String includedUsers, String excludedUsers, Integer maxArchiveRetries, Boolean block, String rollCycle,
+                               Long maxLogSize, Integer maxQueueWeight, String archiveCommand) throws ConfigurationException, IllegalStateException;
+
```

Will fix now

File: src/java/org/apache/cassandra/db/ColumnFamilyStoreMBean.java (hunk 3 above)
Comment (ekaterinadimitrova2) [reply src/java/org/apache/cassandra/db/ColumnFamilyStoreMBean.java:57]: Yeah, I think we can mark as @Deprecated the old ones, without actually removing them, and add also new ones

Comment (dcapwell) [reply src/java/org/apache/cassandra/db/ColumnFamilyStoreMBean.java:57]: just pushed a `v2`.  We don't have the following `fn(range.toString()) == range`, so went with a "simpler" `token:token` string

Comment (adelapena) [reply src/java/org/apache/cassandra/db/ColumnFamilyStoreMBean.java:57]: Looks good. I guess that the `v2` part of the name is to avoid overloaded methods. Since the method is applied to multiple token ranges, maybe we can call it `forceCompactionForTokenRanges`, in plural, so we don't keep the `*v2` name forever? Feel free to ignore if you don't agree.

File: src/java/org/apache/cassandra/db/ColumnFamilyStore.java [hunk 8]
```diff
@@ -2332,11 +2333,33 @@ public void forceMajorCompaction(boolean splitOutput)
...
+    }
+
+    static Set<Range<Token>> toTokenRanges(IPartitioner partitioner, String... strings)
+    {
+        Token.TokenFactory tokenFactory = partitioner.getTokenFactory();
+        Set<Range<Token>> tokenRanges = new HashSet<>();
+        for (String str : strings)
+        {
+            String[] splits = str.split(TOKEN_DELIMITER);
+            assert splits.length == 2 : String.format("Unable to parse token range %s; needs to have two tokens seperated by %s", str, TOKEN_DELIMITER);
```
Comment (adelapena) [inline src/java/org/apache/cassandra/db/ColumnFamilyStore.java]: Nit: typo
```suggestion
            assert splits.length == 2 : String.format("Unable to parse token range %s; needs to have two tokens separated by %s", str, TOKEN_DELIMITER);
```

File: test/unit/org/apache/cassandra/db/ColumnFamilyStoreMBeanTest.java [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+        testToTokenRanges(OrderPreservingPartitioner.instance);
+    }
+
+    @Test
+    public void testToTokenRangesByteOrderedPartitioner()
+    {
+        testToTokenRanges(ByteOrderedPartitioner.instance);
+    }
+
+    private static void testToTokenRanges(IPartitioner partitioner)
```
Comment (adelapena) [inline test/unit/org/apache/cassandra/db/ColumnFamilyStoreMBeanTest.java:97]: Nit: we could include a check for the parsing error:
```java
Assertions.assertThatThrownBy(() -> ColumnFamilyStore.toTokenRanges(partitioner, "invalid"))
          .isInstanceOf(AssertionError.class)
          .hasMessageContaining("Unable to parse token range invalid; needs to have two tokens separated by :");
```

File: src/java/org/apache/cassandra/db/ColumnFamilyStoreMBean.java (hunk 3 above)
Comment (dcapwell) [reply src/java/org/apache/cassandra/db/ColumnFamilyStoreMBean.java:57]: went with `s`

File: test/unit/org/apache/cassandra/db/ColumnFamilyStoreMBeanTest.java [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.cassandra.db;
+
+import java.util.Arrays;
+import java.util.Random;
+import java.util.Set;
```
Comment (adelapena) [inline test/unit/org/apache/cassandra/db/ColumnFamilyStoreMBeanTest.java]: Nit: unused import

File: test/unit/org/apache/cassandra/db/ColumnFamilyStoreMBeanTest.java [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+    }
+
+    private static void testToTokenRanges(IPartitioner partitioner)
+    {
+        Token.TokenFactory tokenFactory = partitioner.getTokenFactory();
+        Gen<Token> tokenGen = tokenGen(partitioner);
+        qt().forAll(tokenGen, tokenGen)
+            .checkAssert((left, right) ->
+                         assertThat(ColumnFamilyStore.toTokenRanges(partitioner, toString(tokenFactory, left, right)))
+                         .isEqualTo(ImmutableSet.of(new Range(left, right))));
```
Comment (adelapena) [inline test/unit/org/apache/cassandra/db/ColumnFamilyStoreMBeanTest.java]:
```suggestion
                         .isEqualTo(ImmutableSet.of(new Range<>(left, right))));
```

File: test/unit/org/apache/cassandra/db/ColumnFamilyStoreMBeanTest.java [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+        return tokenFactory.toString(left) + ColumnFamilyStore.TOKEN_DELIMITER + tokenFactory.toString(right);
+    }
+
+    private static Gen<Token> tokenGen(IPartitioner partitioner)
+    {
+        // Random and RandomSource can not share the same seed, but there is a workaround...
+        // use RandomSource to generate a seed!
+        return rs -> partitioner.getRandomToken(new Random(rs.next(Constraint.none())));
+    }
+}
```
Comment (adelapena) [inline test/unit/org/apache/cassandra/db/ColumnFamilyStoreMBeanTest.java]: Nit: missed newline at end of file

**Review by adelapena — APPROVED** [review]
Changes look good to me. I have left a couple of trivial nits that can be addressed on commit. +1 if CI looks good.
